using-python-for-science/understanding-errors.md

```diff
@@ -0,0 +1,562 @@
+much faster at that third step, and thus speed up your work dramatically. In
+our view, what distinguishes experienced programmers from beginners is less the
+ability to get things right from the get-go, but the ability to quickly recover
+from errors.
+
+Let's see how Python represents errors, by typing in some obviously-wrong code.
+
+```python
+empty_list = []
+empty_list[5]  # attempt to access the 6th element
+```
+
+Running these lines in your Python terminal will cause the following to be printed to your screen:
+
+```pytb
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+IndexError: list index out of range
+```
+
+So we see that when we try to *index* something incorrectly (this can be a
+list, but also a tuple, a NumPy array, or many other custom objects that
+support indexing), Python raises an IndexError. This is one of the most common
+errors you'll see. Typically, it's caused because the list/tuple/etc that you
+are indexing is shorter than expected, and you need to figure out why that is.
+
+Similarly, if you try to access a *dictionary* item that doesn't exist, you get
+a similar error called a KeyError:
+
+```python
+empty_dict = {}
+empty_dict['hello']
+```
+
+raises:
+
+```pytb
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+KeyError: 'hello'
+```
+
+### A fast tour of the most common Python errors
+
+Writing code that is not correct Python (all the above was correct Python, just with unexpected inputs) will raise a SyntaxError. For example, the program control keywords in Python (if, for, def, and others) are reserved and cannot be assigned to, since otherwise programs would not work after the point at which these keywords are redefined. Therefore, it is a SyntaxError to try to redefine these keywords:
+
+```python
+if = 5
+```
+
+raises
+
+```pytb
+  File "<stdin>", line 1
+    if = 5
+       ^
+SyntaxError: invalid syntax
+```
+
+One particularly tricky form of SyntaxError occurs when you forget to close a
+parenthesis. That's because it's valid Python to open a parenthesis on one line
+and continue whatever goes inside the parenthesis along the continuing lines.
+It's only when you write something that *couldn't* go in the parentheses that
+you get the syntax error. So, if you get a SyntaxError and you can't for the
+life of you figure out why the printed line is wrong, look up: you might have
+forgotten to close a parenthesis a bit higher up.
+
+```python
+array_data = (
+    (1, 0, 20),
+    (0, 1, 10),
+    (0, 0,  1),
+
+def transform_data(image, tf_matrix):
+    pass
+```
+
+which raises:
+
+```pytb
+  File "<stdin>", line 6
+    def transform_data(image, tf_matrix):
+    ^
+SyntaxError: invalid syntax
+```
+
+There is nothing wrong with that transform_data definition --- but we did
+forget to close the second parenthesis on the array_data definition! So again,
+if you're completely baffled about why Python is giving you a SyntaxError,
+don't forget to look up!
```

Comment:
Maybe remind the readers that if they use flake8, those should be picked up by the static linter (maybe in a footnote?)

Reply:
Yup, change added, will push shortly